java/arcs/sdk/Reference.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+@file:Suppress("FunctionName") // Allow pseudo-constructors.
+
+package arcs.sdk
+
+import arcs.core.data.RawEntity
+import arcs.sdk.common.ReferenceId
+import arcs.sdk.storage.StorageKey
+
+/** Represents a reference (ie. pointer) to an [Entity]. */
+typealias Reference = arcs.core.storage.Reference
+
+/**
+ * Creates a new [Reference] to an [Entity] with the given [entityId] within the region identified
+ * by [regionKey].
+ */
+fun Reference(entityId: ReferenceId, regionKey: StorageKey) =
+    Reference(entityId as arcs.core.common.ReferenceId, regionKey, version = null)
+
+/**
+ * Dereferences a [Reference] into a [RawEntity] or `null` if the [Reference] is invalid.
+ */
+@ExperimentalReferenceApi
+suspend fun Reference.dereferenceRaw(): RawEntity? = TODO("not implemented")
+
+/**
+ * Dereferences a [Reference] into an [Entity] of type [T], or `null` if the [Reference] is invalid.
+ */
+@ExperimentalReferenceApi
+suspend fun <T : Entity> Reference.dereference(entitySpec: EntitySpec<T>): T? =
+    dereferenceRaw()?.let {
+        TODO("Use the entitySpec to convert the rawEntity into the actual entity")
+    }
```

Comment:
Same question with these. I imagine that developers will be handed a fully specced reference and will just call .dereference() on it with no args. Maybe these should live in core? I also hope we don't need to expose RawEntity in SDK.

Reply:
SGMT

Reply:
Since these are labeled Experimental, I'm going to leave them here for now - and see where the winds take the rest of the reference work.

Reply:
Since these are labeled Experimental, I'm going to leave them here for now - and see where the winds take the rest of the reference work.

Reply:
The schema2kotlin Kotlin generator should be supplying the entity spec I should think. Couldn't we have something like this:

```
class MyEntity : Entity {
   val bar_ref: Reference
   val bar: EntityReference<BarEntity> get() = { () -> bar_ref.dereference(BarEntity_Spec()) }
}
```

The idea being, you'd write myEntity.bar.get()

EntityReference would just be a SAM lambda with a get() method that invokes entitySpec.deserialize(dereferenceRaw())